drivers/ethernet/phy/phy_mii.c

```diff
@@ -388,9 +403,12 @@ static int phy_mii_cfg_link(const struct device *dev,
 		return -ENODEV;
 	}
 
+	k_sem_take(&data->sem, K_FOREVER);
```

Comment:
After this point, there are many early returns which do not give back the semaphore. This can result in a deadlock.